### PR TITLE
Increasing resistance to fault injection attacks related with memory operations.

### PR DESCRIPTION
**Pull request by piotr-now**
Signed-off-by: Piotr Nowicki <piotr.nowicki@arm.com>

## Description
This PR adds flow monitor to the function mbedtls_platform_memcmp() and a start form a random location in function uECC_vli_equal() to increase resistance to fault injection

## Status
**READY**

## Requires Backporting
Yes

## Todos
- [ ] Tests
- [ ] Backported

### CODE REVIEW

**Comment by piotr-now** [comment]
Force push to repair CI failures

**Comment by shelib01** [comment]
@piotr-now , please add function description for memory operation functions.
Also specify in comments why  regular memset may be used in  mbedtls_platform_memcpy

**Comment by piotr-now** [comment]
> @piotr-now , please add function description for memory operation functions.
> Also specify in comments why regular memset may be used in mbedtls_platform_memcpy

I added some more description as a comments in source code, also for mentioned memset. As for the function descriptions themselves, they are all available in the header files.

File: tinycrypt/ecc.c [hunk 1]
```diff
@@ -289,17 +289,29 @@ uECC_word_t uECC_vli_equal(const uECC_word_t *left, const uECC_word_t *right)
 	uECC_word_t tmp1, tmp2;
 	volatile int i;
 
-	for (i = NUM_ECC_WORDS - 1; i >= 0; --i) {
+	/* Start from a random location and check the correct number of iterations */
+	int flow_monitor = 0;
+	int start_offset = 1 + mbedtls_platform_random_in_range(NUM_ECC_WORDS - 1);
```
Comment (AndrzejKurek) [inline tinycrypt/ecc.c:294]: Why not just `mbedtls_platform_random_in_range(NUM_ECC_WORDS)`?

File: tinycrypt/ecc.c [hunk 2]
```diff
@@ -289,17 +289,29 @@ uECC_word_t uECC_vli_equal(const uECC_word_t *left, const uECC_word_t *right)
...
 	}
 
-	/* i should be -1 now */
+	/* Random delay to increase security */
 	mbedtls_platform_random_delay();
-	diff |= i ^ -1;
 
-	return diff;
+	/* Return 0 only when diff is 0 and flow_counter is equal to NUM_ECC_WORDS */
+	return (diff | (flow_monitor ^ NUM_ECC_WORDS));
```
Comment (AndrzejKurek) [inline tinycrypt/ecc.c:314]: Shouldn't there be a cast to "uECC_word_t" here?

File: tinycrypt/ecc.c (hunk 1 above)
Comment (piotr-now) [reply tinycrypt/ecc.c:294]: NUM_ECC_WORDS is equal to 8. The probability of falling 0, i.e. performing operations without starting from a random location is quite high. In this way the random value will be always in the range [1, 7], i.e. starting from a random location will be guaranteed.

File: tinycrypt/ecc.c (hunk 2 above)
Comment (piotr-now) [reply tinycrypt/ecc.c:315]: Yes, It should be. I changed the type of flow_monitor to"uECC_word_t"

**Comment by piotr-now** [comment]
I did force push with fixed cast founded by Andrzej

File: tinycrypt/ecc.c (hunk 1 above)
Comment (AndrzejKurek) [reply tinycrypt/ecc.c:294]: 0 is also a random location :)

**Comment by AndrzejKurek** [comment]
Please investigate the CI failure and rerun the test if you think that it is a spurious failure.

**Comment by piotr-now** [comment]
> Please investigate the CI failure and rerun the test if you think that it is a spurious failure.

This CI failure comes from the ssl-opt.sh script. This is a known CI bug and not related to changes.

The more important tests **continuous-integration/jenkins/pr-merge** and **PR-3336-merge TLS Testing** are Pass

File: tinycrypt/ecc.c (hunk 1 above)
Comment (piotr-now) [reply tinycrypt/ecc.c]: If you say so, I changed it

**Comment by piotr-now** [comment]
One more force push after Andrzej's suggestion about starting from a 0 random location

**Comment by shelib01** [comment]
@piotr-now , can we merge it ?

**Comment by piotr-now** [comment]
> @piotr-now , can we merge it ?

Yes, I just merged it

**Pull request merged**